datafusion/core/src/datasource/file_format/csv.rs

```diff
@@ -236,13 +236,15 @@ impl FileFormat for CsvFormat {
 
     async fn create_physical_plan(
         &self,
-        _state: &SessionState,
+        state: &SessionState,
         conf: FileScanConfig,
         _filters: Option<&Arc<dyn PhysicalExpr>>,
     ) -> Result<Arc<dyn ExecutionPlan>> {
         let exec = CsvExec::new(
             conf,
-            self.options.has_header,
+            self.options
+                .has_header
+                .unwrap_or(state.config().options().catalog.has_header),
```

Comment:
It took me a while to understand why this is reading from `catalog.has_header` (which is different than `format.has_header`). I think it is because catalog.has_header is the default value that is used when has_header has not been explicitly specified for the table.

If so this make sense - but maybe it is worth a comment on `catalog.has_header` explaining this (I can make a follow on PR too)